cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java

```diff
@@ -0,0 +1,446 @@
+      // if it had AuthEnforce annotation.
+      mv = new MethodVisitor(Opcodes.ASM5, mv) {
+
+        boolean hasAuthEnforce;
+        AuthAnnotation authAnnotation;
+        final Map<String, Integer> paramAnnotation = new HashMap<>();
+
+        @Override
+        public AnnotationVisitor visitAnnotation(String desc, boolean visible) {
+          AnnotationVisitor annotationVisitor = super.visitAnnotation(desc, visible);
+
+
+          // If the annotation is not visible or its not AuthEnforce just skip
+          if (!(visible && AuthEnforce.class.getName().equals(Type.getType(desc).getClassName()))) {
+            return annotationVisitor;
+          }
+
+          // AuthEnforce annotation is present so visit it
+          // set the flag which will be used to decide whether to visit parameter annotation or not
+          hasAuthEnforce = true;
+          final Type[] enforceOn = new Type[1];
+          final List<String> entities = new ArrayList<>();
+          final Set<Action> actions = new HashSet<>();
+
+          // We found AuthEnforce annotation so visit the annotation to collect all the specified details
+          // See AnnotationVisitor documentation to know the order in which the below overridden methods are
+          // called and their responsibilities.
+          annotationVisitor = new AnnotationVisitor(Opcodes.ASM5, annotationVisitor) {
+            // gets the entity id class on which enforcement has to performed.
+            @Override
+            public void visit(String name, Object value) {
+
+              enforceOn[0] = (Type) value;
+
+              super.visit(name, value);
+            }
+
+            // gets the entity parts and actions for enforcement by visiting the array for values
+            @Override
+            public AnnotationVisitor visitArray(String name) {
+              AnnotationVisitor av = super.visitArray(name);
+              return new AnnotationVisitor(Opcodes.ASM5, av) {
+                // get the entity parts on which enforcement needs to be performed
+                @Override
+                public void visit(String name, Object value) {
+                  entities.add((String) value);
+                  super.visit(name, value);
+                }
+
+                //  get the actions which needs to be checked during enforcement
+                @Override
+                public void visitEnum(String name, String desc, String value) {
+                  actions.add(Action.valueOf(value));
+                  super.visitEnum(name, desc, value);
+                }
+              };
+            }
+          };
+          authAnnotation = new AuthAnnotation(entities, enforceOn[0], actions);
+          return annotationVisitor;
+        }
+
+        @Override
+        public AnnotationVisitor visitParameterAnnotation(final int parameter, String desc, boolean visible) {
+          AnnotationVisitor annotationVisitor = super.visitParameterAnnotation(parameter, desc, visible);
```

Comment:
Similar to above, I believe the code would be simpler if you use `AnnotationNode` as well.